Filter migration nodes based on VMI pod's node selectors

martindekov opened the pull request:
Filtering nodes for migration based on VMI pod's nodeSelection field. We construct label selector based on pod's node selector when picking up available nodes for migration

#### Problem:
Currently we omit filtering of nodes based on criteria like whether node is schedulable, what is the CPU model of the node and what is the CPU which the VM requires if any. We don't respect VMI's pod nodeSelectors when figuring out compatible nodes for VM migration.

#### Solution:
When listing possible nodes for migration we use label selector built using the virtual machine instance pod's nodeSelectors.

#### Related Issue(s):
https://github.com/harvester/harvester/issues/8797

#### Test plan:
Unit tests - added new nodes and new test case with the fields

Basic E2E test validation on virtualization:
1. Create 2 node harvester cluster `node-1` and `node-2`
2. Create VM - `vm-1` on the cluster
3. By default since nodes are similar - created on virtualization you should be able to see the other node and be able to migrate to it
4. Adding affinity rules to vm should make it unschedulable, and then once you label the nodes the vm will be scheduled and each node which contains the labels should be seen as migration target as well

Validate Node Scheduling works:
1. Create a VM
2. By default VM is on - `Run virtual machine on any available node` so migration button was present
3. Enabled VM to be hosted only on the current node - `Run virtual machine on specific node - (Live migration is not supported)`
4. Migration button dissapeared
5. Reverted VM to original state of - `Run virtual machine on any available node`
6. Migrated VM to another node
7. After migration enabled - `Run virtual machine on specific node - (Live migration is not supported)`
8. Migration button dissapeared

Results for Basic E2E test validation on virtualization:

<details>
<summary>Built cluster with this change and custom image to replace the controller and used it on 2 node cluster</summary>
<br>

```
node-1:/home/rancher # kubectl get deployment harvester -n harvester-system -o yaml | grep image
          value: '{"repository":"rancher/support-bundle-kit","tag":"master-head","imagePullPolicy":"IfNotPresent"}'
        image: rancher/harvester:cpufilter8797-head-1
        imagePullPolicy: IfNotPresent
```
</details>

<details>
<summary>Virtual machine on one node allows migration to another node</summary>
<br>
<img width="2288" height="655" alt="image" src="https://github.com/user-attachments/assets/6be07461-20b8-46a7-a72c-d21851fb39dd" />
<img width="2288" height="850" alt="image" src="https://github.com/user-attachments/assets/ccfc483c-8f1a-4a78-9695-e3887a74f657" />
</details>



<details>
<summary>Adding affinity while vm is on node-1 makes node-2 dissapear as potential target</summary>
<br>
<img width="2288" height="850" alt="image" src="https://github.com/user-attachments/assets/ad532033-9866-49e9-b85c-3109c6449741" />
<img width="2559" height="505" alt="image" src="https://github.com/user-attachments/assets/2af36874-a2b8-4b28-aad4-f7980f2be249" />
<img width="2559" height="505" alt="image" src="https://github.com/user-attachments/assets/d082b989-ac0d-4a74-9e03-65bebc817dc5" />
<img width="2559" height="794" alt="image" src="https://github.com/user-attachments/assets/3de50fe7-37ff-44ba-8d4c-7322afaae2bf" />
<img width="2559" height="794" alt="image" src="https://github.com/user-attachments/assets/5f073800-8493-4585-a401-943f83e89f60" />
<img width="2559" height="794" alt="image" src="https://github.com/user-attachments/assets/2e79ed9f-27b9-45c5-b061-03e14c0acf16" />
</details>

<details>
<summary>While migration is running you migration button dissapears</summary>
<br>
<img width="2288" height="850" alt="image" src="https://github.com/user-attachments/assets/461f8980-9121-4695-8d9d-1057de9aaee6" />
</details>

Results for Validate Node Scheduling works:
<details>
<summary>Create VM by default schedulable to all nodes so migratable</summary>
<br>

<img width="1134" height="552" alt="image" src="https://github.com/user-attachments/assets/aee3104b-e231-4482-b0ef-6c3408b69e56" />
<img width="1638" height="491" alt="image" src="https://github.com/user-attachments/assets/70d58ccc-d0ef-4f83-9cb9-6a19279eea8a" />

```
node-1:/home/rancher # kubectl get pod virt-launcher-vm-1-rrgzp -n default -o jsonpath='{.spec.nodeSelector}' | jq
{
  "kubernetes.io/arch": "amd64",
  "kubevirt.io/schedulable": "true",
  "machine-type.node.kubevirt.io/q35": "true"
}
```
</details>


<details>
<summary>Make VM schedulable on specific node - current one and migration button dissapears</summary>
<br>
<img width="1283" height="430" alt="image" src="https://github.com/user-attachments/assets/a7eea831-275b-48cf-9596-a9cc20989037" />
<img width="2274" height="731" alt="image" src="https://github.com/user-attachments/assets/63ff2548-509e-44c2-a728-c44c8bee309e" />

```
node-1:/home/rancher # kubectl get pod virt-launcher-vm-1-cbjpd -n default -o jsonpath='{.spec.nodeSelector}' | jq
{
  "kubernetes.io/arch": "amd64",
  "kubernetes.io/hostname": "node-2",
  "kubevirt.io/schedulable": "true",
  "machine-type.node.kubevirt.io/q35": "true"
}
```
</details>

<details>
<summary>Revert vm schedulable to all nodes, migrate it and see migration back and forth is allowed</summary>
<br>
<img width="1646" height="480" alt="image" src="https://github.com/user-attachments/assets/165fb57e-1881-41ca-badd-90cc66688ded" />

```
node-1:/home/rancher # kubectl get pod virt-launcher-vm-1-cbx4r -n default -o jsonpath='{.spec.nodeSelector}' | jq
{
  "cpu-feature.node.kubevirt.io/arch-capabilities": "true",
  "cpu-feature.node.kubevirt.io/cmp_legacy": "true",
  "cpu-feature.node.kubevirt.io/flush-l1d": "true",
  "cpu-feature.node.kubevirt.io/flushbyasid": "true",
  "cpu-feature.node.kubevirt.io/gds-no": "true",
  "cpu-feature.node.kubevirt.io/hypervisor": "true",
  "cpu-feature.node.kubevirt.io/ibpb-brtype": "true",
  "cpu-feature.node.kubevirt.io/lbrv": "true",
  "cpu-feature.node.kubevirt.io/mds-no": "true",
  "cpu-feature.node.kubevirt.io/overflow-recov": "true",
  "cpu-feature.node.kubevirt.io/pause-filter": "true",
  "cpu-feature.node.kubevirt.io/perfmon-v2": "true",
  "cpu-feature.node.kubevirt.io/pfthreshold": "true",
  "cpu-feature.node.kubevirt.io/pschange-mc-no": "true",
  "cpu-feature.node.kubevirt.io/rdctl-no": "true",
  "cpu-feature.node.kubevirt.io/rfds-no": "true",
  "cpu-feature.node.kubevirt.io/sbpb": "true",
  "cpu-feature.node.kubevirt.io/skip-l1dfl-vmentry": "true",
  "cpu-feature.node.kubevirt.io/ssbd": "true",
  "cpu-feature.node.kubevirt.io/stibp": "true",
  "cpu-feature.node.kubevirt.io/succor": "true",
  "cpu-feature.node.kubevirt.io/tsc-deadline": "true",
  "cpu-feature.node.kubevirt.io/tsc-scale": "true",
  "cpu-feature.node.kubevirt.io/tsc_adjust": "true",
  "cpu-feature.node.kubevirt.io/vgif": "true",
  "cpu-feature.node.kubevirt.io/virt-ssbd": "true",
  "cpu-feature.node.kubevirt.io/vmcb-clean": "true",
  "cpu-feature.node.kubevirt.io/x2apic": "true",
  "cpu-model-migration.node.kubevirt.io/EPYC-Genoa": "true",
  "kubernetes.io/arch": "amd64",
  "kubernetes.io/hostname": "node-1",
  "kubevirt.io/schedulable": "true",
  "machine-type.node.kubevirt.io/pc-q35-9.2": "true"
}
node-1:/home/rancher # kubectl get pod virt-launcher-vm-1-dwjrf -n default -o jsonpath='{.spec.nodeSelector}' | jq
{
  "kubernetes.io/arch": "amd64",
  "kubevirt.io/schedulable": "true",
  "machine-type.node.kubevirt.io/q35": "true"
}
```
</details>

<details>
<summary>Enabling vm to be ran on specific node after migration removes ability to migrate</summary>
<br>
<img width="1275" height="690" alt="image" src="https://github.com/user-attachments/assets/23f5419a-ee02-4445-8c8d-f275ac13ef6f" />
<img width="2291" height="728" alt="image" src="https://github.com/user-attachments/assets/df7370d6-1d30-42bb-aff4-1bba5dd9fa54" />

```
node-1:/home/rancher # kubectl get pods
NAME                       READY   STATUS    RESTARTS   AGE
virt-launcher-vm-1-nxqn4   2/2     Running   0          2m23s
node-1:/home/rancher # kubectl get pod virt-launcher-vm-1-nxqn4 -n default -o jsonpath='{.spec.nodeSelector}' | jq
{
  "kubernetes.io/arch": "amd64",
  "kubernetes.io/hostname": "node-1",
  "kubevirt.io/schedulable": "true",
  "machine-type.node.kubevirt.io/q35": "true"
}
```
</details>

#### Additional documentation or context
N/A

martindekov committed:
Add node selectors for CPU based on vmi spec

Adding the following selectors based on vmi:
* kubevirtv1.CPUModelLabel
* kubevirtv1.CPUFeatureLabel
* kubevirtv1.NodeSchedulable

The way we construct them is based off from the
virt-controller implementation using those constants
go construct the selectors against the nodes.

Also now we don't assume virtual machines are schedulable
on all nodes in case affinities are not set so removed
early return and we require node to be schedulable before
adding it as migration option.

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov committed:
Add default CPU model

Adding default CPU model check in case the VMI is missing
setting it through the spec

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov commented:
Thanks for the input @ibrokethecloud added in the latest commmit - when model is missing, look up default one, and if set use the selector from it, fixed the tests as well

ihcsim commented on pkg/api/vm/handler.go:
```diff
@@ -622,25 +622,75 @@ func isDrained(node *corev1.Node) bool {
-		return labels.Everything(), nil
+	nodeSelector := labels.NewSelector()
+
+	if vmi != nil && vmi.Spec.Affinity != nil && vmi.Spec.Affinity.NodeAffinity != nil && vmi.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
```
The CPU model and feature labels are added to the `virt-launcher` pod's `.spec.nodeSelector`, not the VMI's node affinity. Ultimately, it's the pod that gets scheduled to the node, not the VMI.

martindekov commented on pkg/api/vm/handler.go:
```diff
@@ -622,25 +622,75 @@ func isDrained(node *corev1.Node) bool {
+	}
+
+	if cpuModel != "" {
+		cpuModelKey := fmt.Sprintf("%s%s", kubevirtv1.CPUModelLabel, cpuModel)
```
I think this suggestion:
https://github.com/harvester/harvester/pull/9372#discussion_r2610476531
Would solve this concern as well, let me know

Yu-Jack replied:
I think you can just copy the original node selector of the pod.

martindekov committed:
Use vmi pod's nodeSelectors as label selectors

Using VMI's latest pod nodeSelectors on the current
node as label selecor for other nodes. The change
includes the following:
* new method to take latest VMI pod on current node where
the VM actually runs
* removed affinity logic
* adding pod cache when registering vmAtionHandler
* simply taking pod's nodeSelectors and create label selectors
for maching nodes for migration
* Changed unit tests and added pod cache fake to support the new logic
* Renamed nodeSelector as nodeFilter to differentiate the nodeSelector
from the pod and the label selector node filter that we create

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

Yu-Jack commented on pkg/api/vm/handler.go:
```diff
@@ -622,23 +624,46 @@ func isDrained(node *corev1.Node) bool {
 		}
+		nodeFilter = nodeFilter.Add(*requirement)
 	}
+	return nodeFilter, nil
```
I just tested and realized that we should skip the `"kubernetes.io/hostname"`. Otherwise, it always returns the node where it's from.

For example:

This is pod node selector.

```yaml
nodeSelector:
  cpu-feature.node.kubevirt.io/amd-ssbd: "true"
  cpu-feature.node.kubevirt.io/amd-stibp: "true"
  cpu-feature.node.kubevirt.io/flush-l1d: "true"
  cpu-feature.node.kubevirt.io/gds-no: "true"
  cpu-feature.node.kubevirt.io/hypervisor: "true"
  cpu-feature.node.kubevirt.io/ibpb: "true"
  cpu-feature.node.kubevirt.io/ibrs: "true"
  cpu-feature.node.kubevirt.io/md-clear: "true"
  cpu-feature.node.kubevirt.io/movdir64b: "true"
  cpu-feature.node.kubevirt.io/movdiri: "true"
  cpu-feature.node.kubevirt.io/ss: "true"
  cpu-feature.node.kubevirt.io/stibp: "true"
  cpu-feature.node.kubevirt.io/tsc_adjust: "true"
  cpu-feature.node.kubevirt.io/vmx: "true"
  cpu-feature.node.kubevirt.io/vmx-activity-wait-sipi: "true"
  cpu-feature.node.kubevirt.io/vmx-tsc-scaling: "true"
  cpu-model-migration.node.kubevirt.io/SierraForest: "true"
  kubernetes.io/arch: amd64
  kubernetes.io/hostname: harvester2
  kubevirt.io/schedulable: "true"
  machine-type.node.kubevirt.io/pc-q35-9.2: "true"
```

When I was trying to migrate it, the option list was empty because it also used `kubernetes.io/hostname` as a filter.

This video did two things

- Created a new VM and migrated it.
- Migrated it again. -> The list was empty.

https://github.com/user-attachments/assets/1ea9fbb8-3218-4cc2-8819-a9e4b9cf3c56

martindekov replied:
Rebased and skipped this key `kubernetes.io/hostname`, you can resolve if the latest commit makes sense regarding this feedback

martindekov committed:
Skip hostname label

Skipping hostname label selector and adding unit test
to validate this

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

ibrokethecloud commented:
in general the PR looks good however we need to fine tune it a bit more, the current implementation drops the original check where we calculate VMI node affinity criteria

https://github.com/harvester/harvester/pull/9372/files#diff-6a02f49ff2d9e428071b3a2d7c39eea15917d24b5e67997a714a7cb6535eec09L625

The VMI node affinity gets mapped to launcher pod's nodeAffinity.

We need to apply both the node affinity and node selectors to ensure correct filtering is applied to identify target nodes.

For example consider a sample VM which has a user defined node scheduling criteria

<img width="2025" height="467" alt="image" src="https://github.com/user-attachments/assets/074e7ea9-726b-4850-884e-f8357dea8ffb" />

This maps to `.spec.template.spec.affinity.nodeAffinity` field

```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  annotations:
    harvesterhci.io/mac-address: '{"default":"b2:e1:ff:3f:52:ed"}'
    harvesterhci.io/vmRunStrategy: RerunOnFailure
    harvesterhci.io/volumeClaimTemplates: '[{"metadata":{"name":"migration-test-zone-disk-0-vymfh","annotations":{"harvesterhci.io/imageId":"default/image-6sb88"}},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"10Gi"}},"volumeMode":"Block","storageClassName":"longhorn-image-6sb88"}}]'
    kubevirt.io/latest-observed-api-version: v1
    kubevirt.io/storage-observed-api-version: v1
    network.harvesterhci.io/ips: '[]'
  creationTimestamp: "2026-01-15T23:20:39Z"
  finalizers:
  - kubevirt.io/virtualMachineControllerFinalize
  - wrangler.cattle.io/VMController.CleanupPVCAndSnapshot
  generation: 1
  labels:
    harvesterhci.io/creator: harvester
    harvesterhci.io/os: linux
  name: migration-test-zone
  namespace: default
  resourceVersion: "1781131"
  uid: 7a16cb99-d38a-47e3-a3ab-445f6b060ea8
spec:
  runStrategy: RerunOnFailure
  template:
    metadata:
      annotations:
        harvesterhci.io/sshNames: '[]'
      creationTimestamp: null
      labels:
        harvesterhci.io/vmName: migration-test-zone
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: zone
                operator: In
                values:
                - dell
 ```

which maps to the pods `.spec.affinity.nodeAffinity` field

```
k get pod virt-launcher-migration-test-zone-7gj5v -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: 83f8996d591a2a827c882584dbec0636357107406481de17c831664472ec4174
    cni.projectcalico.org/podIP: 10.52.1.60/32
    cni.projectcalico.org/podIPs: 10.52.1.60/32
    descheduler.alpha.kubernetes.io/request-evict-only: ""
    harvesterhci.io/sshNames: '[]'
    istio.io/reroute-virtual-interfaces: k6t-eth0
    k8s.v1.cni.cncf.io/network-status: |-
      [{
          "name": "k8s-pod-network",
          "ips": [
              "10.52.1.60"
          ],
          "default": true,
          "dns": {}
      }]
    kubectl.kubernetes.io/default-container: compute
    kubevirt.io/domain: migration-test-zone
    kubevirt.io/migrationTransportUnix: "true"
    kubevirt.io/vm-generation: "1"
    post.hook.backup.velero.io/command: '["/usr/bin/virt-freezer", "--unfreeze", "--name",
      "migration-test-zone", "--namespace", "default"]'
    post.hook.backup.velero.io/container: compute
    pre.hook.backup.velero.io/command: '["/usr/bin/virt-freezer", "--freeze", "--name",
      "migration-test-zone", "--namespace", "default"]'
    pre.hook.backup.velero.io/container: compute
    traffic.sidecar.istio.io/kubevirtInterfaces: k6t-eth0
  creationTimestamp: "2026-01-15T23:20:39Z"
  generateName: virt-launcher-migration-test-zone-
  generation: 1
  labels:
    harvesterhci.io/vmName: migration-test-zone
    kubevirt.io: virt-launcher
    kubevirt.io/created-by: 0265a7d1-94a3-4c2c-96f8-73be31b99749
    kubevirt.io/nodeName: dell-140-tink-system
    vm.kubevirt.io/name: migration-test-zone
  name: virt-launcher-migration-test-zone-7gj5v
  namespace: default
  ownerReferences:
  - apiVersion: kubevirt.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: VirtualMachineInstance
    name: migration-test-zone
    uid: 0265a7d1-94a3-4c2c-96f8-73be31b99749
  resourceVersion: "1780934"
  uid: c6139f5d-df45-420d-ae46-8e68ff975cdd
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: zone
            operator: In
            values:
            - dell
          - key: node-labeller.kubevirt.io/obsolete-host-model
            operator: DoesNotExist
```

In the current implementation we ignore this and only use the labels from `.spec.nodeSelector`.

This will result in nodes not meeting the original VM scheduling criterial being available in the list of nodes available from migration.

if would be nice to also have a unit test that covers this scenario.

martindekov committed:
Revert affinity rules requirement from actual state

Reverting usage of affinity rules, but now instead of
using the rules defined from the desired state in
the vmi spec, we use the actual state from the virt
launcher pod.

Adding unit tests as well and refactor the original
code to reduce nesting.

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov committed:
Remove indexer and allow only 1 VMI pod

Removing usage of indexer along with the indexer
and replaced with list with selector. We now rely
on status field of the pod to tell us the active
pods and error out in the following cases:
* 0 pods - no VM is actually running so cannot
 determine node matches
* >1 pods - multiple vmi pods are running which means
migration is currently happening

and allow the following case:
* exactly 1 pod is currently in non Failed or
Succeeded state

that way we know migration is not happening and pod
is in other viable states - Pending/Running/Unknown

Fixed fakes and unit tests to accomodate the changes
above

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov committed:
Count pods as opposed to relying on vmi status field

Counting pods in the desired state as opposed to relying
on activePods status field. After migration a vmi can
still have multiple "active" pods, even though only 1
pod is actually containing the running virtual machine.

This commit changes the logic so now we list all pods
and count how much are running and allow migration
only in case only 1 is running.

Fixing the unit tests to accomodate the change as well.

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov committed:
Fix linter

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov commented:
@w13915984028 @ibrokethecloud @ihcsim @Yu-Jack Team if you can please do a second review, updated testing done section where I validated:
1. Migration does not appear while migration is running
2. Affinity is respected 

Attached pictures and explained basic flow of the change. Let me know your thoughts

w13915984028 commented on pkg/api/vm/handler.go:
```diff
@@ -637,23 +638,120 @@ func isDrained(node *corev1.Node) bool {
+	if err != nil {
+		return labels.Everything(), err
+	}
+	if latestPod == nil {
```
per definition of getVMIPod, the `latestPod == nil` seems not possible, after the above error case processing

martindekov replied:
Good catch Jian, removed the check and renamed to vmiPod I think it suits better

martindekov committed:
Remove latestPod nil check and rename to vmiPod

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

ibrokethecloud commented:
lgtm apart from the minor nit. if we find hostname in node selector then VM is not live migratable and we should return no valid nodes matching the migration criteria.

martindekov committed:
Address feedback from Ivan

Signed-off-by: Martin Dekov <martin.dekov@suse.com>
